app/src/main/java/com/cdio/solitaire/controller/GameStateController.kt

```diff
@@ -0,0 +1,316 @@
+        else if (stack.tail!!.rank.ordinal == 0) stack.tail
+        else null
+    }
+
+    /**
+     * moveStack moves a stack and adds cardToUpdate information to the move object.
+     */
+    private fun moveStack( // TODO: Does this actually need cardsToMove? See popStack comment.
+        sourceStack: CardStack,
+        cardsToMove: Int,
+        targetStack: CardStack
+    ): Card? {
+        targetStack.pushStack(sourceStack.popStack(cardsToMove)!!)
+        return cardToUpdate(sourceStack)
+    }
+
+    private fun moveCard(sourceStack: CardStack, targetStack: CardStack): Card? {
+        return moveStack(sourceStack, 1, targetStack)
+    }
+
+    /**
+     * performMove is a fork function on moveType of the move object.
+     * It maps each moveType to their respective methods.
+     */
+    fun performMove(move: Move) {
+        var cardToUpdate: Card? = null
+        when (move.moveType) {
+            MoveType.MOVE_STACK -> cardToUpdate = moveStack(
+                move.sourceStack!!,
+                getCardPosition(move.sourceCard, move.sourceStack),
+                move.targetStack!!
+            )
+            MoveType.MOVE_FROM_FOUNDATION,
+            MoveType.MOVE_FROM_TALON,
+            MoveType.MOVE_TO_FOUNDATION -> cardToUpdate =
+                moveCard(move.sourceStack!!, move.targetStack!!)
+            MoveType.FLIP_TALON -> flipTalon()
+            MoveType.DRAW_STOCK -> cardToUpdate = drawFromStock()
+            else -> Log.e("MoveTypeNotDefined", "This cannot happen.")
+        }
+        move.cardToUpdate = cardToUpdate
+        gameState.moves.add(move)
+    }
+
+    /**
+     * Searches for the position of a card in a given cardStack, measured from the tail element.
+     */
+    private fun getCardPosition(card: Card?, stack: CardStack): Int {
+        if (card!!.stackID != stack.stackID) {
+            Log.e("CardNotInStack", "Card stackID differs from stack stackID")
+            return -1
+        }
+        var stackCard = stack.tail!!
+        var i = 1
+        while (stackCard.rank.ordinal != card.rank.ordinal) {
+            i++
+            stackCard = stackCard.prev!!
+        }
+        return i
+    }
+
+    fun getLastMove(): Move {
+        return gameState.moves.last()
+    }
+
+    fun updateSortedDeck(card: Card) {
+        sortedDeck[(card.suit.ordinal - 1) * 13 + card.rank.ordinal - 1] = card
+    }
+
+    /**
+     * Used to ensure consistency in tableauOrdering when performing moves.
+     */
+    private fun tableauOrdering(card: Card, targetStack: CardStack): Boolean =
+        if (targetStack.stackID !in 1..7) false
+        else when (targetStack.size) {
+            0 -> card.rank.ordinal == 13 // Only kings can go on an empty tableau
+            else -> {
+                val targetCard = targetStack.tail!!
+                val ranksMatch = targetCard.rank.ordinal - card.rank.ordinal == 1
+                val offColor = targetCard.suit.isRed() xor card.suit.isRed()
```

Comment:
I think using bitwise comparison of two booleans is a bit over engineered, when a simple direct comparison does the job, and it makes the code more readable. 🙂 
```suggestion
                val offColor = targetCard.suit.isRed() != card.suit.isRed()
```

Reply:
Granted, but funny. 🤣 Changed isRed() to offSuit() in the Suit enum class.